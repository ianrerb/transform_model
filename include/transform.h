#ifndef TRANSFORM_H
#define TRANSFORM_H
#define PI 3.14159265358979
#include"pricemodel.h"
#include<complex.h>
#include<fftw3.h>
#include<math.h>
#include<vector>

struct Option {double strike; double premium; }; //struct to hold strike, premium pairs

class transform_base {//abc
  public:
    transform_base(){ };
    virtual ~transform_base(){ };
    
    virtual std::vector<Option> Prices(const double &Spot, const double &Strike, const double &C, const pricemodel &model) const =0;
 };
 
 class FFT : public transform_base {
  public:
    FFT(unsigned int N_, double alpha_,double eta_) : N(N_), alpha(alpha_), eta(eta_){ }; 
    ~FFT(){ }; 
    
    void setN(const unsigned int &val){ N = val; };
    void setAlpha(const double &val){ alpha = val; };
    void setEta(const double &val){eta = val; };
    
    unsigned int getN() const { return N; };
    double Alpha() const { return alpha; };
    double Eta() const { return eta; };
    double Lambda() const {return 2.0*PI/(static_cast<double>(N)*eta); };
     
    std::vector<Option> Prices(const double &Spot, const double &Strike, const double &C, const pricemodel &model) const; //defined in FFTengine.cpp
    

  private:
    double alpha, eta;
    unsigned int N;
    
    void ComputeX(double complex X[], const double &Spot, const double &Strike, const double &C, const pricemodel &model) const;  //defined in FFTengine.cpp
  };

#endif
