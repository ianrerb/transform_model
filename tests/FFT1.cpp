#include"../include/transform.h"
#include<cmath>
#include<iostream>
#include<vector>

using namespace std;

int main(){
  double Spot, Strike, rate, q, T, vol;
  Spot = 100;
  Strike = 90;
  rate = .05;
  q = 0.0;
  T=1.0;
  vol = .3;

  GBM model(vol,rate,q,T);
  FFT Engine(0,0,0);

  vector<double> alpha, eta;
  vector<unsigned int> n; 
  
  alpha.push_back(.01);
  alpha.push_back(.5);
  alpha.push_back(1);
  alpha.push_back(1.5);
  alpha.push_back(2);
  alpha.push_back(5);
  alpha.push_back(10);

  eta.push_back(.15);
  eta.push_back(.1);
  eta.push_back(.05);
  
  n.push_back(64);
  n.push_back(256);
  n.push_back(1024);

  for(vector<double>::iterator ait = alpha.begin(); ait!=alpha.end(); ait++){
    for(vector<double>::iterator eit = eta.begin(); eit!= eta.end(); eit++){
      for(vector<unsigned int>::iterator nit = n.begin(); nit!=n.end(); nit++){
	  Engine.setN(*nit);
	  Engine.setAlpha(*ait);
	  Engine.setEta(*eit);
	  cout<<Engine.Prices(Spot,Strike,exp(-rate*T),model)[*nit/2].premium<<"  ";
	}
      }
    cout<<endl;
    }
  return 0;
  }
