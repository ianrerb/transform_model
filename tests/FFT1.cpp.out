211.907  211.91  211.91  134.048  134.054  134.054  60.4539  59.4812  59.4812  
19.6922  19.6974  19.6974  19.6264  19.6974  19.6974  19.7932  19.6973  19.6974  
19.6911  19.6974  19.6974  19.5632  19.6974  19.6974  18.7819  19.6974  19.6974  
19.6919  19.6974  19.6974  19.5162  19.6974  19.6974  17.8268  19.6975  19.6974  
19.695  19.6974  19.6974  19.4991  19.6974  19.6974  17.0632  19.6976  19.6974  
19.702  19.6974  19.6974  20.324  19.6974  19.6974  19.7178  19.6968  19.6974  
20.9053  19.6974  19.6974  2.42137  19.6974  19.6974  96.43  19.6688  19.6974  
